packages/@sanity/migrate/src/runner/utils/batchMutations.ts

```diff
@@ -0,0 +1,55 @@
+import {Mutation as SanityMutation} from '@sanity/client'
+import arrify from 'arrify'
+
+// We're working on "raw" mutations, e.g what will be put into the mutations array in the request body
+const PADDING_SIZE = '{"mutations":[]}'.length
+
+/**
+ *
+ * @param mutations - Async iterable of either single values or arrays of values
+ * @param maxBatchSize - Max batch size in bytes
+ * Todo: add support for transaction ids too
+ */
+export async function* batchMutations(
+  mutations: AsyncIterableIterator<SanityMutation | SanityMutation[]>,
+  maxBatchSize: number,
+  options?: {preserveTransactions: boolean},
+): AsyncIterableIterator<SanityMutation | SanityMutation[]> {
+  let currentBatch: SanityMutation[] = []
+  let currentBatchSize = 0
+
+  for await (const mutation of mutations) {
+    if (options?.preserveTransactions && Array.isArray(mutation)) {
+      yield currentBatch
+      yield mutation
+      currentBatch = []
+      currentBatchSize = 0
+      continue
+    }
+
+    // the mutation itself may exceed the payload size, need to handle that
```

Comment:
Will we throw an error in this scenario?

Reply:
I think it's better to try posting the mutation. The limits we currently have in the migraiton runner is quite conservative compared to the payload size accepted by the mutate endpoint (which is 4MB)